Nonblocking Collectives (Draft) (#456)
openshmem-org/specification: content/shmem_broadcast_nb.tex

      The type of \dest{} should match that implied in the SYNOPSIS section.}
\apiargument{IN}{source}{Symmetric address of the source data object.
      The type of \source{} should match that implied in the SYNOPSIS section.}
\apiargument{IN}{nelems}{
  The number of elements in \source{} and \dest{} arrays.
  For \FUNC{shmem\_broadcastmem\_nb}, elements are bytes;
  for \FUNC{shmem\_broadcast\{32,64\}\_nb}, elements are 4 or 8 bytes,
  respectively.
}
\apiargument{IN}{PE\_root}{Zero-based ordinal of the \ac{PE}, with respect to
    the team, from which the data is copied.}
\apiargument{OUT}{request}{An opaque request handle identifying the collective
operation.}

\end{apiarguments}

\apidescription{   
    \openshmem nonblocking broadcast routines are collective routines over a 
    valid \openshmem team.
    They copy the \source{} data object on the \ac{PE} specified by
    \VAR{PE\_root} to the \dest{} data object on the \acp{PE}
    participating in the collective operation.
    The same \dest{} and \source{} data objects and the same value of
    \VAR{PE\_root} must be passed by all \acp{PE} participating in the
    collective operation.

>     A call to the nonblocking broadcast routine posts the operation and returns
>     immediately without necessarily completing the operation. On the successful
>     post of the operation, an opaque request handle is created and returned. The

Reviewer: ```suggestion
    A call to the nonblocking broadcast routine initiates the operation and returns
    immediately without necessarily completing the operation. On success,
    an opaque request handle is created and returned. The
```

    \FUNC{shmem\_req\_wait}. When the operation is complete, the request handle
    is deallocated and cannot be reused.

    Like blocking broadcast, before any \ac{PE} calls a broadcast routine, the following
    conditions must be ensured:
    \begin{itemize}
    \item The \dest{} array on all \acp{PE} participating in the broadcast
      is ready to accept the broadcast data.
    \item All \acp{PE} in the \VAR{team} argument must participate in
      the operation.
    \item If the \VAR{team} compares equal to \LibConstRef{SHMEM\_TEAM\_INVALID} or is
      otherwise invalid, the behavior is undefined.
    \item \ac{PE} numbering is relative to the team. The specified
      root \ac{PE} must be a valid \ac{PE} number for the team,
      between \CONST{0} and \VAR{N$-$1}, where \VAR{N} is the size of
      the team.
    \end{itemize}
    Otherwise, the behavior is undefined.

    Upon completion of a nonblocking broadcast routine, the following are true for the local
    \ac{PE}:
    \begin{itemize}
    \item The \dest{} data object is
      updated.
    \item The \source{} data object may be safely reused.
    \end{itemize}
}

\apireturnvalues{

>   Zero on successful posting of the collective
>   operation; otherwise, nonzero.

Reviewer: ```suggestion
  Zero on success and nonzero otherwise.
```
Copy/paste from elsewhere in the spec

    the team, from which the data is copied.}
\apiargument{IN}{tag}{A user defined tag to order the collective operation;
SHMEM\_COLL\_UNORDERED can be provided if no order is required.}
\apiargument{OUT}{request}{An opaque request handle identifying the collective
operation.}

\end{apiarguments}

\apidescription{   
    \openshmem nonblocking broadcast routines are collective routines over a 
    valid \openshmem team.
    They copy the \source{} data object on the \ac{PE} specified by
    \VAR{PE\_root} to the \dest{} data object on the \acp{PE}
    participating in the collective operation.
    The same \dest{} and \source{} data objects and the same value of
    \VAR{PE\_root} must be passed by all \acp{PE} participating in the
    collective operation.

    A call to the nonblocking broadcast routine posts the operation and returns
    immediately without necessarily completing the operation. On the successful
    post of the operation, an opaque request handle is created and returned. The
    operation is completed after a call to \FUNC{shmem\_req\_test} or
    \FUNC{shmem\_req\_wait}. When the operation is complete, the request handle

>     is deallocated and cannot be reused.

Reviewer 2: To my understanding. This is currently worded in a way that both `shmem_req_test` and `shmem_req_wait` guarantee completion.

I believe the correct language exists in the intro on lines 17 and 18.

      otherwise invalid, the behavior is undefined.
    \item \ac{PE} numbering is relative to the team. The specified
      root \ac{PE} must be a valid \ac{PE} number for the team,
      between \CONST{0} and \VAR{N$-$1}, where \VAR{N} is the size of
      the team.
    \end{itemize}
    Otherwise, the behavior is undefined.

    Upon completion of a nonblocking broadcast routine, the following are true for the local
    \ac{PE}:
    \begin{itemize}
    \item The \dest{} data object is updated.
    \item If the local \ac{PE} is \VAR{PE\_root}, the data has been copied
          out of the \source{} data object.
    \end{itemize}
}

\apireturnvalues{
  Zero on success and nonzero otherwise.
}

\apinotes{
    Team handle error checking and integer return codes are currently undefined.
    Implementations may define these behaviors as needed, but programs should
    ensure portability by doing their own checks for invalid team handles and for
    \LibConstRef{SHMEM\_TEAM\_INVALID}.
}

\begin{apiexamples}

\apicexample
    {In the following \Cstd[11] example, the call to multiple \FUNC{shmem\_broadcast\_nb} copies the \source{}

>     params on \ac{PE} $0\dots npes-1$ to the \dest{} params on \acp{PE} $0\dots npes-1$. This effectively performs a \FUNC{shmem\_collect} operation using nonblocking broadcasts.}

Reviewer 3: ```suggestion
    params on \ac{PE} $0\dots npes-1$ to the \dest{} params on \acp{PE} $0\dots npes-1$. This effectively performs a \FUNC{shmem\_fcollect} operation using nonblocking broadcasts.}
```

>     params on \ac{PE} $0\dots npes-1$ to the \dest{} params on \acp{PE} $0\dots npes-1$. This effectively performs a \FUNC{shmem\_collect} operation using nonblocking broadcasts.}

Reviewer 4: \ac{PE} -> \acp{PE}

>     params on \ac{PE} $0\dots npes-1$ to the \dest{} params on \acp{PE} $0\dots npes-1$. This effectively performs a \FUNC{shmem\_collect} operation using nonblocking broadcasts.}

Reviewer 4: shmem_collect -> shmem_fcollect
